Content.Client/DoAfter/DoAfterSystem.cs

```diff
@@ -7,217 +7,180 @@
-            {
-                Remove(component, doAfter);
-            }
-
-            foreach (var doAfter in state.DoAfters)
-            {
-                if (component.DoAfters.ContainsKey(doAfter.ID))
-                    continue;
+            if (component.DoAfters.ContainsKey(doAfter.ID))
+                continue;
 
-                component.DoAfters.Add(doAfter.ID, doAfter);
-            }
+            component.DoAfters.Add(doAfter.ID, doAfter);
         }
+    }
 
-        private void OnCancelledDoAfter(CancelledDoAfterMessage ev)
-        {
-            if (!TryComp<DoAfterComponent>(ev.Uid, out var doAfter))
-                return;
-
-            Cancel(doAfter, ev.ID);
-        }
+    private void OnCancelledDoAfter(CancelledDoAfterMessage ev)
+    {
+        if (!TryComp<DoAfterComponent>(ev.Uid, out var doAfter))
+            return;
 
-        /// <summary>
-        ///     Remove a DoAfter without showing a cancellation graphic.
-        /// </summary>
-        public void Remove(DoAfterComponent component, ClientDoAfter clientDoAfter)
-        {
-            component.DoAfters.Remove(clientDoAfter.ID);
+        Cancel(doAfter, ev.ID);
+    }
 
-            var found = false;
+    /// <summary>
+    ///     Remove a DoAfter without showing a cancellation graphic.
+    /// </summary>
+    public void Remove(DoAfterComponent component, Shared.DoAfter.DoAfter doAfter, bool found = false)
+    {
+        component.DoAfters.Remove(doAfter.ID);
+        component.CancelledDoAfters.Remove(doAfter.ID);
+    }
 
-            component.CancelledDoAfters.Remove(clientDoAfter.ID);
+    /// <summary>
+    ///     Mark a DoAfter as cancelled and show a cancellation graphic.
+    /// </summary>
+    ///     Actual removal is handled by DoAfterEntitySystem.
+    public void Cancel(DoAfterComponent component, byte id)
+    {
+        if (component.CancelledDoAfters.ContainsKey(id))
+            return;
 
-            if (!found)
-                component.DoAfters.Remove(clientDoAfter.ID);
-        }
+        if (!component.DoAfters.ContainsKey(id))
+            return;
 
-        /// <summary>
-        ///     Mark a DoAfter as cancelled and show a cancellation graphic.
-        /// </summary>
-        ///     Actual removal is handled by DoAfterEntitySystem.
-        public void Cancel(DoAfterComponent component, byte id)
-        {
-            if (component.CancelledDoAfters.ContainsKey(id))
-                return;
+        var doAfterMessage = component.DoAfters[id];
+        doAfterMessage.Cancelled = true;
+        doAfterMessage.CancelledTime = _gameTiming.CurTime;
+        component.CancelledDoAfters.Add(id, doAfterMessage);
+    }
 
-            if (!component.DoAfters.ContainsKey(id))
-                return;
+    // TODO separate DoAfter & ActiveDoAfter components for the entity query.
+    public override void Update(float frameTime)
+    {
+        if (!_gameTiming.IsFirstTimePredicted)
+            return;
```

Comment:
Is this supposed to not be running shared code?

Reply:
Next PR, since translating it over is going to take some time. Right now it works as is.